presto-main/src/main/java/com/facebook/presto/sql/planner/RowExpressionInterpreter.java

```diff
@@ -0,0 +1,781 @@
+        this.expression = requireNonNull(expression, "expression is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.session = requireNonNull(session, "session is null");
+        this.optimize = optimize;
+        this.functionInvoker = new InterpretedFunctionInvoker(metadata.getFunctionManager());
+        this.determinismEvaluator = new com.facebook.presto.sql.relational.DeterminismEvaluator(metadata.getFunctionManager());
+
+        this.visitor = new Visitor();
+    }
+
+    public Type getType()
+    {
+        return expression.getType();
+    }
+
+    public Object evaluate()
+    {
+        checkState(!optimize, "evaluate() not allowed for optimizer");
+        return expression.accept(visitor, new NoPagePositionContext());
+    }
+
+    public Object evaluate(int position, Page page)
+    {
+        checkState(!optimize, "evaluate(int, Page) not allowed for optimizer");
+        return expression.accept(visitor, new SinglePagePositionContext(position, page));
+    }
+
+    public Object optimize(SymbolResolver inputs)
+    {
+        checkState(optimize, "evaluate(SymbolResolver) not allowed for interpreter");
+        return expression.accept(visitor, inputs);
+    }
+
+    private class Visitor
+            implements RowExpressionVisitor<Object, Object>
+    {
+        @Override
+        public Object visitInputReference(InputReferenceExpression node, Object context)
+        {
+            throw new UnsupportedOperationException("Field references not supported in interpreter");
+        }
+
+        @Override
+        public Object visitConstant(ConstantExpression node, Object context)
+        {
+            // Unify output type if necessary
+            if (node.getType().getJavaType() == Slice.class && node.getValue() instanceof String) {
+                return utf8Slice((String) node.getValue());
+            }
+            return node.getValue();
+        }
+
+        @Override
+        public Object visitVariableReference(VariableReferenceExpression node, Object context)
+        {
+            if (context instanceof SymbolResolver) {
+                return ((SymbolResolver) context).getValue(new Symbol(node.getName()));
+            }
+            return node;
+        }
+
+        @Override
+        public Object visitCall(CallExpression node, Object context)
+        {
+            List<Type> argumentTypes = new ArrayList<>();
+            List<Object> argumentValues = new ArrayList<>();
+            for (RowExpression expression : node.getArguments()) {
+                Object value = expression.accept(this, context);
+                argumentValues.add(value);
+                argumentTypes.add(expression.getType());
+            }
+
+            FunctionHandle functionHandle = node.getFunctionHandle();
+
+            // Special casing for large constant array construction
+            if (functionHandle.getSignature().getName().toUpperCase().equals(ARRAY_CONSTRUCTOR)) {
```

Comment:
Maybe move all special case handling to separate functions. That would make this function easier to read. Especially for like, the logic is too long it's hard to match parenthesis.